File: modules/mpv-player/ios/NativePlayer/NativePlayerModule.swift
import ExpoModulesCore

// Exceptions shared by both platforms (tvOS builds this file too — the
// generated ExpoModulesProvider references the class on all Apple platforms).

internal final class NativePlayerUnsupportedException: Exception, @unchecked Sendable {
	override var reason: String {
		"The native player is only available on iOS (iPhone/iPad)"
	}
}

internal final class PlayerAlreadyPresentedException: Exception, @unchecked Sendable {
	override var reason: String {
		"A native player is already presented; call load() to swap the stream instead"
	}
}

internal final class NoActivePlayerException: Exception, @unchecked Sendable {
	override var reason: String {
		"No native player is currently presented"
	}
}

internal final class InvalidStreamUrlException: Exception, @unchecked Sendable {
	override var reason: String {
		"The stream config contains an unparseable URL"
	}
}

/// Module-level (non-view) player: presents a full-screen native
/// UIViewController with SwiftUI controls over the shared MPVPlayerEngine.
/// All Jellyfin orchestration (stream negotiation, session reporting,
/// next-episode resolution) stays in JS — see providers/NativePlayerProvider.tsx.
/// JS must attach every event listener BEFORE calling presentPlayer: module
/// events emitted with no listener are dropped.
public class NativePlayerModule: Module {
	#if os(iOS)
	private var session: NativePlayerSession?
	#endif

	public func definition() -> ModuleDefinition {
		Name("NativePlayer")

		Events(
			"onLoad", "onProgress", "onPlaybackStateChange", "onError",
			"onTracksReady", "onPictureInPictureChange",
			"onTrackSelectionRequested", "onSpeedChange",
			"onQualitySelected", "onSubtitleScaleChange",
			"onOrientationChangeRequested",
			"onNextEpisodeRequested", "onPreviousEpisodeRequested",
			"onEpisodeSelected", "onPlaybackEnded", "onDismiss",
			"onSubtitleSearchRequested", "onSubtitleDownloadRequested"
		)

		// MARK: - Lifecycle

		AsyncFunction("presentPlayer") { (config: PlayerPresentConfigRecord, promise: Promise) in
			#if os(iOS)
			if self.session != nil {
				promise.reject(PlayerAlreadyPresentedException())
				return
			}
			guard config.stream.toVideoLoadConfig() != nil else {
				promise.reject(InvalidStreamUrlException())
				return
			}
			let session = NativePlayerSession(
				emit: { [weak self] name, payload in
					self?.sendEvent(name, payload)
				},
				onTornDown: { [weak self] in
					self?.session = nil
				},
				presenterProvider: { [weak self] in
					self?.appContext?.utilities?.currentViewController()
				}
			)
			self.session = session
			do {
				try session.present(config: config, promise: promise)
			} catch {
				self.session = nil
				promise.reject(error)
			}
			#else
			promise.reject(NativePlayerUnsupportedException())
			#endif
		}.runOnQueue(.main)

		AsyncFunction("load") { (config: PlayerPresentConfigRecord, promise: Promise) in
			#if os(iOS)
			guard let session = self.session else {
				promise.reject(NoActivePlayerException())
				return
			}
			guard config.stream.toVideoLoadConfig() != nil else {
				promise.reject(InvalidStreamUrlException())
				return
			}
			session.load(config: config)
			promise.resolve()
			#else
			promise.reject(NativePlayerUnsupportedException())
			#endif
		}.runOnQueue(.main)

		AsyncFunction("dismiss") { (promise: Promise) in
			#if os(iOS)
			guard let session = self.session else {
				// Already gone — dismissal must be idempotent for the JS side.
				promise.resolve()
				return
			}
			session.dismiss(reason: .programmatic) {
				promise.resolve()
			}
			#else
			promise.resolve()
			#endif
		}.runOnQueue(.main)

		Function("isPresented") { () -> Bool in
			#if os(iOS)
			return self.session != nil
			#else
			return false
			#endif
		}

		// MARK: - Late-arriving data pushes

		AsyncFunction("updateSegments") { (segments: [MediaSegmentRecord]) in
			#if os(iOS)
			self.session?.viewModel.updateSegments(segments)
			#endif
		}.runOnQueue(.main)

		AsyncFunction("updateNextEpisode") { (next: NextEpisodeRecord?) in
			#if os(iOS)
			self.session?.viewModel.updateNextEpisode(next)
			#endif
		}.runOnQueue(.main)

		AsyncFunction("updateChapters") { (chapters: [ChapterRecord]) in
			#if os(iOS)
			self.session?.viewModel.updateChapters(chapters)
			#endif
		}.runOnQueue(.main)

		AsyncFunction("updateTrickplay") { (trickplay: TrickplayRecord?) in
			#if os(iOS)
			self.session?.viewModel.updateTrickplay(trickplay)
			#endif
		}.runOnQueue(.main)

		AsyncFunction("updateTrackMenus") { (menus: TrackMenusRecord) in
			#if os(iOS)
			self.session?.viewModel.updateTrackMenus(menus)
			#endif
		}.runOnQueue(.main)

		AsyncFunction("updateMetadata") { (metadata: MetadataRecord) in
			#if os(iOS)
			self.session?.viewModel.updateMetadata(metadata)
			self.session?.applyNowPlayingMetadata(metadata)
			#endif
		}.runOnQueue(.main)

		AsyncFunction("updateEpisodeList") { (episodes: [EpisodeListItemRecord]) in
			#if os(iOS)
			self.session?.viewModel.updateEpisodeList(episodes)
			#endif
		}.runOnQueue(.main)

		AsyncFunction("updateSubtitleSearch") { (state: SubtitleSearchStateRecord) in
			#if os(iOS)
			self.session?.viewModel.updateSubtitleSearch(state)
			#endif
		}.runOnQueue(.main)

		// Adds a sidecar subtitle to the live mpv handle and selects it
		// (client-side downloaded subtitle file, e.g. OpenSubtitles fallback).
		AsyncFunction("addExternalSubtitle") { (url: String) in
			#if os(iOS)
			self.session?.engine.addSubtitleFile(url: url, select: true)
			#endif
		}.runOnQueue(.main)

		// MARK: - Transport (driven by the JS coordinator: WebSocket remote
		// control commands from the Jellyfin server, not the native UI)

		AsyncFunction("play") {
			#if os(iOS)
			self.session?.engine.play()
			#endif
		}.runOnQueue(.main)

		AsyncFunction("pause") {
			#if os(iOS)
			self.session?.engine.pause()
			#endif
		}.runOnQueue(.main)

		AsyncFunction("seekTo") { (positionSec: Double) in
			#if os(iOS)
			self.session?.engine.seekTo(position: positionSec)
			#endif
		}.runOnQueue(.main)

		AsyncFunction("setSpeed") { (speed: Double) in
			#if os(iOS)
			self.session?.viewModel.setSpeed(speed)
			#endif
		}.runOnQueue(.main)

		AsyncFunction("getCurrentPosition") { () -> Double in
			#if os(iOS)
			return self.session?.engine.getCurrentPosition() ?? 0
			#else
			return 0
			#endif
		}.runOnQueue(.main)

		AsyncFunction("getDuration") { () -> Double in
			#if os(iOS)
			return self.session?.engine.getDuration() ?? 0
			#else
			return 0
			#endif
		}.runOnQueue(.main)

		// MARK: - Track plumbing for the JS identity resolver
		// (utils/jellyfin/subtitleUtils.ts drives these through a
		// SubtitleSelectablePlayer facade; selection logic never lives here)

		AsyncFunction("getSubtitleTracks") { () -> [[String: Any]] in
			#if os(iOS)
			return self.session?.engine.getSubtitleTracks() ?? []
			#else
			return []
			#endif
		}.runOnQueue(.main)

		AsyncFunction("setSubtitleTrack") { (mpvId: Int) in
			#if os(iOS)
			self.session?.engine.setSubtitleTrack(mpvId)
			#endif
		}.runOnQueue(.main)

		AsyncFunction("disableSubtitles") {
			#if os(iOS)
			self.session?.engine.disableSubtitles()
			#endif
		}.runOnQueue(.main)

		AsyncFunction("getAudioTracks") { () -> [[String: Any]] in
			#if os(iOS)
			return self.session?.engine.getAudioTracks() ?? []
			#else
			return []
			#endif
		}.runOnQueue(.main)

		AsyncFunction("setAudioTrack") { (mpvId: Int) in
			#if os(iOS)
			self.session?.engine.setAudioTrack(mpvId)
			#endif
		}.runOnQueue(.main)

		AsyncFunction("getTechnicalInfo") { () -> [String: Any] in
			#if os(iOS)
			return self.session?.engine.getTechnicalInfo() ?? [:]
			#else
			return [:]
			#endif
		}.runOnQueue(.main)

		OnDestroy {
			#if os(iOS)
			// App reload / module teardown while a player is up: tear down
			// synchronously so the mpv handle and audio session are released.
			self.session?.teardownImmediately()
			self.session = nil
			#endif
		}
	}
}

File: modules/mpv-player/android/src/main/java/expo/modules/mpvplayer/nativeplayer/NativePlayerModule.kt
package expo.modules.mpvplayer.nativeplayer

import android.os.Handler
import android.os.HandlerThread
import expo.modules.kotlin.Promise
import expo.modules.kotlin.exception.CodedException
import expo.modules.kotlin.functions.Queues
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

class PlayerAlreadyPresentedException :
    CodedException("A native player is already presented; call load() to swap the stream instead")

class NoActivePlayerException :
    CodedException("No native player is currently presented")

class InvalidStreamUrlException :
    CodedException("The stream config contains an unparseable URL")

class NativePlayerModule : Module() {

    private var session: NativePlayerSession? = null

    // HandlerThread for off-main MPV property reads to prevent UI thread blockage
    private val propsThread = HandlerThread("mpv-props").apply { start() }
    private val propsHandler = Handler(propsThread.looper)

    override fun definition() = ModuleDefinition {
        Name("NativePlayer")

        Events(
            "onLoad", "onProgress", "onPlaybackStateChange", "onError",
            "onTracksReady", "onPictureInPictureChange",
            "onTrackSelectionRequested", "onSpeedChange",
            "onQualitySelected", "onSubtitleScaleChange",
            "onOrientationChangeRequested",
            "onNextEpisodeRequested", "onPreviousEpisodeRequested",
            "onEpisodeSelected", "onPlaybackEnded", "onDismiss",
            "onSubtitleSearchRequested", "onSubtitleDownloadRequested",
            "onMuteStateChanged"
        )

        // MARK: - Lifecycle

        AsyncFunction("presentPlayer") { config: PlayerPresentConfigRecord, promise: Promise ->
            if (this@NativePlayerModule.session != null) {
                promise.reject(PlayerAlreadyPresentedException())
                return@AsyncFunction
            }
            if (config.stream.toVideoLoadConfig() == null) {
                promise.reject(InvalidStreamUrlException())
                return@AsyncFunction
            }

            val newSession = NativePlayerSession(
                emit = { name, payload ->
                    sendEvent(name, payload)
                },
                onTornDown = {
                    this@NativePlayerModule.session = null
                },
                presenterProvider = {
                    appContext.currentActivity
                }
            )
            this@NativePlayerModule.session = newSession

            try {
                newSession.present(config, promise)
            } catch (e: Exception) {
                this@NativePlayerModule.session = null
                promise.reject("PRESENT_FAILED", e.message, e)
            }
        }

        AsyncFunction("load") { config: PlayerPresentConfigRecord, promise: Promise ->
            val s = this@NativePlayerModule.session
            if (s == null) {
                promise.reject(NoActivePlayerException())
                return@AsyncFunction
            }
            if (config.stream.toVideoLoadConfig() == null) {
                promise.reject(InvalidStreamUrlException())
                return@AsyncFunction
            }
            s.load(config)
            promise.resolve(null)
        }

        AsyncFunction("dismiss") { promise: Promise ->
            val s = this@NativePlayerModule.session
            if (s == null) {
                promise.resolve(null)
                return@AsyncFunction
            }
            s.dismiss(reason = "programmatic") {
                promise.resolve(null)
            }
        }

        Function("isPresented") {
            this@NativePlayerModule.session != null
        }

        // MARK: - Late-Arriving Data Pushes

        AsyncFunction("updateSegments") { segments: List<MediaSegmentRecord> ->
            this@NativePlayerModule.session?.viewModel?.updateSegments(segments)
        }

        AsyncFunction("updateNextEpisode") { next: NextEpisodeRecord? ->
            this@NativePlayerModule.session?.viewModel?.updateNextEpisode(next)
        }

        AsyncFunction("updateChapters") { chapters: List<ChapterRecord> ->
            this@NativePlayerModule.session?.viewModel?.updateChapters(chapters)
        }

        AsyncFunction("updateTrickplay") { trickplay: TrickplayRecord? ->
            this@NativePlayerModule.session?.viewModel?.updateTrickplay(trickplay)
        }

        AsyncFunction("updateTrackMenus") { menus: TrackMenusRecord ->
            this@NativePlayerModule.session?.viewModel?.updateTrackMenus(menus)
        }

        AsyncFunction("updateMetadata") { metadata: MetadataRecord ->
            this@NativePlayerModule.session?.viewModel?.updateMetadata(metadata)
        }

        AsyncFunction("updateEpisodeList") { episodes: List<EpisodeListItemRecord> ->
            this@NativePlayerModule.session?.viewModel?.updateEpisodeList(episodes)
        }

        AsyncFunction("updateSubtitleSearch") { state: SubtitleSearchStateRecord ->
            this@NativePlayerModule.session?.viewModel?.updateSubtitleSearch(state)
        }

        // Transient one-line notice on the player's own notice surface, for
        // things the JS coordinator decides (automatic subtitles on mute).
        AsyncFunction("showNotice") { text: String ->
            this@NativePlayerModule.session?.viewModel?.showNotice(text)
        }

        // Jellyfin remote ToggleMute. Same entry point as the TV mute button.
        AsyncFunction("toggleMute") {
            this@NativePlayerModule.session?.viewModel?.toggleMute()
        }

        AsyncFunction("addExternalSubtitle") { url: String ->
            this@NativePlayerModule.session?.renderer?.addSubtitleFile(url, select = true)
        }

        // MARK: - Transport

        AsyncFunction("play") {
            this@NativePlayerModule.session?.viewModel?.play()
        }

        AsyncFunction("pause") {
            this@NativePlayerModule.session?.viewModel?.pause()
        }

        // Main queue, as on iOS: viewModel.seekTo syncs the window's
        // keep-screen-on flag, which only the main thread may touch. On the
        // default background queue that threw mid-call, after the mpv seek but
        // before the tick that moves JS's tracked position, so a backward
        // remote seek followed by an exit reported the pre-seek position.
        AsyncFunction("seekTo") { positionSec: Double ->
            this@NativePlayerModule.session?.viewModel?.seekTo(positionSec)
        }.runOnQueue(Queues.MAIN)

        AsyncFunction("setSpeed") { speed: Double ->
            this@NativePlayerModule.session?.viewModel?.setSpeed(speed)
        }

        AsyncFunction("getCurrentPosition") {
            this@NativePlayerModule.session?.viewModel?.displayPosition ?: 0.0
        }

        AsyncFunction("getDuration") {
            this@NativePlayerModule.session?.viewModel?.duration ?: 0.0
        }

        // MARK: - Track plumbing & technical info (executed off-main)

        AsyncFunction("getSubtitleTracks") { promise: Promise ->
            val r = this@NativePlayerModule.session?.renderer
            if (r == null) {
                promise.resolve(emptyList<Map<String, Any>>())
                return@AsyncFunction
            }
            propsHandler.post {
                val tracks = r.getSubtitleTracks()
                promise.resolve(tracks)
            }
        }

        AsyncFunction("setSubtitleTrack") { mpvId: Int ->
            this@NativePlayerModule.session?.renderer?.setSubtitleTrack(mpvId)
        }

        AsyncFunction("disableSubtitles") {
            this@NativePlayerModule.session?.renderer?.disableSubtitles()
        }

        AsyncFunction("getAudioTracks") { promise: Promise ->
            val r = this@NativePlayerModule.session?.renderer
            if (r == null) {
                promise.resolve(emptyList<Map<String, Any>>())
                return@AsyncFunction
            }
            propsHandler.post {
                val tracks = r.getAudioTracks()
                promise.resolve(tracks)
            }
        }

        AsyncFunction("setAudioTrack") { mpvId: Int ->
            this@NativePlayerModule.session?.renderer?.setAudioTrack(mpvId)
        }

        AsyncFunction("getTechnicalInfo") { promise: Promise ->
            val r = this@NativePlayerModule.session?.renderer
            if (r == null) {
                promise.resolve(emptyMap<String, Any>())
                return@AsyncFunction
            }
            propsHandler.post {
                val info = r.getTechnicalInfo()
                promise.resolve(info)
            }
        }

        OnDestroy {
            this@NativePlayerModule.session?.teardownImmediately()
            this@NativePlayerModule.session = null
            propsThread.quitSafely()
        }
    }
}
